File: ios/VisionCameraMLkitTextRecognitionPlugin.swift
import Foundation
import MLKitCommon
import MLKitTextRecognition
import MLKitTextRecognitionChinese
import MLKitTextRecognitionDevanagari
import MLKitTextRecognitionJapanese
import MLKitTextRecognitionKorean
import MLKitVision
import VisionCamera

@objc(VisionCameraMLkitTextRecognitionPlugin)
public class VisionCameraMLkitTextRecognitionPlugin: FrameProcessorPlugin {
  private var textRecognizer: TextRecognizer
  private var invertColors: Bool = false

  @objc
  public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]! = [:]) {
    let language = options?["language"] as? String ?? "LATIN"

    let textRecognizerOptions: CommonTextRecognizerOptions
    switch language {
    case "LATIN":
      textRecognizerOptions = TextRecognizerOptions()
    case "CHINESE":
      textRecognizerOptions = ChineseTextRecognizerOptions()
    case "DEVANAGARI":
      textRecognizerOptions = DevanagariTextRecognizerOptions()
    case "JAPANESE":
      textRecognizerOptions = JapaneseTextRecognizerOptions()
    case "KOREAN":
      textRecognizerOptions = KoreanTextRecognizerOptions()
    default:
      textRecognizerOptions = TextRecognizerOptions()
    }

    self.textRecognizer = TextRecognizer.textRecognizer(options: textRecognizerOptions)
    self.invertColors = options?["invertColors"] as? Bool ?? false

    super.init(proxy: proxy, options: options)
  }

  public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any
  {
    var map = [String: Any]()

    let image: VisionImage
    if self.invertColors {
      guard
        let invertedImage = VisionCameraMLkitUtils.createInvertedVisionImageFromFrame(frame: frame)
      else {
        print("Failed to invert image colors")
        return [:]
      }
      image = invertedImage
    } else {
      image = VisionCameraMLkitUtils.createVisionImageFromFrame(frame: frame)
    }

    do {
      let text: Text = try self.textRecognizer.results(in: image)
      map["text"] = text.text
      map["blocks"] = createBlocksArray(text.blocks)
    } catch let error {
      print("Text recognition error: \(error.localizedDescription)")
      return [:]
    }

    return map
  }

  private func createElementsArray(_ elements: [TextElement]) -> [[String: Any]] {
    return elements.map { element in
      var map = [String: Any]()
      let bounds = element.frame
      map["bounds"] = VisionCameraMLkitUtils.createBoundsMap(bounds)

      let corners = element.cornerPoints
      map["corners"] = VisionCameraMLkitUtils.createCornersArray(corners)

      map["symbols"] = [:]
      map["confidence"] = nil
      map["angle"] = nil
      map["text"] = element.text
      map["languages"] = element.recognizedLanguages.map { $0.languageCode }
      return map
    }
  }

  private func createLinesArray(_ lines: [TextLine]) -> [[String: Any]] {
    return lines.map { line in
      var map = [String: Any]()
      let bounds = line.frame
      map["bounds"] = VisionCameraMLkitUtils.createBoundsMap(bounds)

      let corners = line.cornerPoints
      map["corners"] = VisionCameraMLkitUtils.createCornersArray(corners)

      map["elements"] = createElementsArray(line.elements)
      map["confidence"] = nil
      map["angle"] = nil
      map["text"] = line.text
      map["languages"] = line.recognizedLanguages.map { $0.languageCode }
      return map
    }
  }

  private func createBlocksArray(_ blocks: [TextBlock]) -> [[String: Any]] {
    return blocks.map { block in
      var map = [String: Any]()
      let bounds = block.frame
      map["bounds"] = VisionCameraMLkitUtils.createBoundsMap(bounds)

      let corners = block.cornerPoints
      map["corners"] = VisionCameraMLkitUtils.createCornersArray(corners)

      map["lines"] = createLinesArray(block.lines)
      map["text"] = block.text
      map["languages"] = block.recognizedLanguages.map { $0.languageCode }
      return map
    }
  }
}

File: android/src/main/java/com/visioncameramlkit/plugins/VisionCameraMLkitTextRecognitionPlugin.kt
package com.visioncameramlkit.plugins

import android.util.Log
import com.facebook.react.bridge.WritableNativeArray
import com.facebook.react.bridge.WritableNativeMap
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.common.MlKitException
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.TextRecognizer
import com.google.mlkit.vision.text.chinese.ChineseTextRecognizerOptions
import com.google.mlkit.vision.text.devanagari.DevanagariTextRecognizerOptions
import com.google.mlkit.vision.text.japanese.JapaneseTextRecognizerOptions
import com.google.mlkit.vision.text.korean.KoreanTextRecognizerOptions
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.mrousavy.camera.core.FrameInvalidError
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy
import com.visioncameramlkit.utils.VisionCameraMLkitUtils.createBoundsMap
import com.visioncameramlkit.utils.VisionCameraMLkitUtils.createCornersArray
import com.visioncameramlkit.utils.VisionCameraMLkitUtils.createInputImageFromFrame
import com.visioncameramlkit.utils.VisionCameraMLkitUtils.createInvertedInputImageFromFrame
import java.util.HashMap

private const val TAG = "VisionCameraMLkitTextRecognitionPlugin"

/**
 * A plugin for recognizing text using ML Kit.
 *
 * @property proxy The VisionCameraProxy instance.
 * @property options The options for the text recognizer.
 */
@Suppress("KDocUnresolvedReference")
class VisionCameraMLkitTextRecognitionPlugin(
    @Suppress(
        "UNUSED_PARAMETER"
    ) proxy: VisionCameraProxy,
    options: Map<String, Any>?
) : FrameProcessorPlugin() {
    private var textRecognizer: TextRecognizer
    private var invertColors: Boolean = false

    init {
        val language = options?.get("language") as? String ?: "LATIN"

        val textRecognizerOptions = when (language) {
            "LATIN" -> TextRecognizerOptions.DEFAULT_OPTIONS
            "CHINESE" -> ChineseTextRecognizerOptions.Builder().build()
            "DEVANAGARI" -> DevanagariTextRecognizerOptions.Builder().build()
            "JAPANESE" -> JapaneseTextRecognizerOptions.Builder().build()
            "KOREAN" -> KoreanTextRecognizerOptions.Builder().build()
            else -> TextRecognizerOptions.DEFAULT_OPTIONS
        }

        this.textRecognizer = TextRecognition.getClient(textRecognizerOptions)
        this.invertColors = options?.get("invertColors") as? Boolean ?: false
    }

    /**
     * Processes the frame and returns the recognized text.
     *
     * @param frame The frame to process.
     * @param arguments The arguments for the frame processor.
     * @return A HashMap of recognized text.
     */
    override fun callback(frame: Frame, arguments: Map<String, Any>?): HashMap<String, Any> {
        try {
            val inputImage = if (this.invertColors) {
                createInvertedInputImageFromFrame(frame)
            } else {
                createInputImageFromFrame(frame)
            }
            val task: Task<Text> = this.textRecognizer.process(inputImage)

            val text: Text? = Tasks.await(task)
            val map = WritableNativeMap()

            map.putString("text", text?.text)
            map.putArray("blocks", createBlocksArray(text?.textBlocks ?: emptyList()))

            return map.toHashMap()
        } catch (e: FrameInvalidError) {
            Log.e(TAG, "FrameInvalidError occurred while processing the frame", e)
        } catch (e: MlKitException) {
            Log.e(TAG, "MlKitException occurred while processing the image with ML Kit", e)
        } catch (@Suppress("TooGenericExceptionCaught") e: Exception) {
            Log.e(TAG, "Unexpected error occurred while recognizing the text", e)
        }

        return hashMapOf()
    }

    /**
     * Creates a WritableNativeArray from a list of symbols.
     *
     * @param symbols The list of symbols.
     * @return A WritableNativeArray of symbols.
     */
    private fun createSymbolsArray(symbols: List<Text.Symbol>): WritableNativeArray {
        return WritableNativeArray().apply {
            symbols.map { symbol ->
                WritableNativeMap().apply {
                    symbol.boundingBox?.let { bounds ->
                        putMap("bounds", createBoundsMap(bounds))
                    }

                    symbol.cornerPoints?.let { corners ->
                        putArray("corners", createCornersArray(corners))
                    }

                    putDouble("confidence", symbol.confidence.toDouble())
                    putDouble("angle", symbol.angle.toDouble())

                    putString("text", symbol.text)
                    putArray("languages", WritableNativeArray().apply { pushString(symbol.recognizedLanguage) })
                }
            }.forEach { pushMap(it) }
        }
    }

    /**
     * Creates a WritableNativeArray from a list of elements.
     *
     * @param elements The list of elements.
     * @return A WritableNativeArray of elements.
     */
    private fun createElementsArray(elements: List<Text.Element>): WritableNativeArray {
        return WritableNativeArray().apply {
            elements.map { element ->
                WritableNativeMap().apply {
                    element.boundingBox?.let { bounds ->
                        putMap("bounds", createBoundsMap(bounds))
                    }

                    element.cornerPoints?.let { corners ->
                        putArray("corners", createCornersArray(corners))
                    }

                    putArray("symbols", createSymbolsArray(element.symbols))

                    putDouble("confidence", element.confidence.toDouble())
                    putDouble("angle", element.angle.toDouble())

                    putString("text", element.text)
                    putArray("languages", WritableNativeArray().apply { pushString(element.recognizedLanguage) })
                }
            }.forEach { pushMap(it) }
        }
    }

    /**
     * Creates a WritableNativeArray from a list of lines.
     *
     * @param lines The list of lines.
     * @return A WritableNativeArray of lines.
     */
    private fun createLinesArray(lines: List<Text.Line>): WritableNativeArray {
        return WritableNativeArray().apply {
            lines.map { line ->
                WritableNativeMap().apply {
                    line.boundingBox?.let { bounds ->
                        putMap("bounds", createBoundsMap(bounds))
                    }

                    line.cornerPoints?.let { corners ->
                        putArray("corners", createCornersArray(corners))
                    }

                    putArray("elements", createElementsArray(line.elements))

                    putDouble("confidence", line.confidence.toDouble())
                    putDouble("angle", line.angle.toDouble())

                    putString("text", line.text)
                    putArray("languages", WritableNativeArray().apply { pushString(line.recognizedLanguage) })
                }
            }.forEach { pushMap(it) }
        }
    }

    /**
     * Creates a WritableNativeArray from a list of text blocks.
     *
     * @param blocks The list of text blocks.
     * @return A WritableNativeArray of text blocks.
     */
    private fun createBlocksArray(blocks: List<Text.TextBlock>): WritableNativeArray {
        return WritableNativeArray().apply {
            blocks.map { block ->
                WritableNativeMap().apply {
                    block.boundingBox?.let { bounds ->
                        putMap("bounds", createBoundsMap(bounds))
                    }

                    block.cornerPoints?.let { corners ->
                        putArray("corners", createCornersArray(corners))
                    }

                    putArray("lines", createLinesArray(block.lines))

                    putString("text", block.text)
                    putArray("languages", WritableNativeArray().apply { pushString(block.recognizedLanguage) })
                }
            }.forEach { pushMap(it) }
        }
    }
}
